### PR TITLE
Crdcdh 462 001 pgu

### PR DESCRIPTION
**Pull request by vshand11**
This PR is for the ticket CRDCDH-462, submission workflow/action.

### CODE REVIEW

File: verifier/submission-verifier.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+const {toPascalCase} = require("../utility/string-util")
+
+function verifySubmissionAction(submissionId, action){ 
+
+    return new SubmissionActionVerifier(submissionId, action);
+}
+
+class SubmissionActionVerifier {
+    constructor(submissionId, action){
+        if(!submissionId) throw Error("submissionID is required!");
```
Comment (knockknockyoo) [inline verifier/submission-verifier.js:11]: Could you declare this error messages in the error-constant.js?

File: verifier/submission-verifier.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            throw new Error(`${ERROR.INVALID_SUBMISSION_NOT_FOUND}, ${this.submissionId }!`);
+        }
+        this.submission = submission[0];
+        return this.submission;
+    }
+
+    isValidAction(actionMaps){
+        let actionMap = actionMaps?.filter((a)=>a.action === toPascalCase(this.action));
+        if(!actionMap || actionMap.length === 0)
+            throw new Error(`Invalid submission action: ${this.action}!`);
```
Comment (knockknockyoo) [inline verifier/submission-verifier.js:29]: Could you declare this error messages in the error-constant.js?

File: verifier/submission-verifier.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        const fromStatus = this.submission.status;
+        if(this.actionMap.fromStatus.indexOf(fromStatus) < 0)
+            throw new Error(`Invalid submission status for the action: ${this.action}!`);
+        this.newStatus = this.actionMap.toStatus;
+    }
+
+    inRoles(userInfo){
+        const role = userInfo?.role;
+        if(this.actionMap.roles.indexOf(role) < 0)
+            throw new Error(`Invalid user role for the action: ${this.action}!`);
```
Comment (knockknockyoo) [inline verifier/submission-verifier.js:41]: Could you declare this error messages in the error-constant.js?

File: verifier/submission-verifier.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    isValidAction(actionMaps){
+        let actionMap = actionMaps?.filter((a)=>a.action === toPascalCase(this.action));
+        if(!actionMap || actionMap.length === 0)
+            throw new Error(`Invalid submission action: ${this.action}!`);
+
+        this.actionMap = actionMap[0];
+        const fromStatus = this.submission.status;
+        if(this.actionMap.fromStatus.indexOf(fromStatus) < 0)
+            throw new Error(`Invalid submission status for the action: ${this.action}!`);
```
Comment (knockknockyoo) [inline verifier/submission-verifier.js:34]: Could you declare this error messages in the error-constant.js?

File: verifier/submission-verifier.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    constructor(submissionId, action){
+        if(!submissionId) throw Error("submissionID is required!");
+        this.submissionId = submissionId;
+        if(!action) throw Error("action is required!");
+        this.action = action;
+    }
+
+    async exists(submissionCollection){
+        const submission = await submissionCollection.find(this.submissionId);
+        if (!submission || submission.length == 0) {
```
Comment (knockknockyoo) [inline verifier/submission-verifier.js:19]: submission.length == 0 should be ===

File: services/submission.js [hunk 6]
```diff
@@ -229,8 +170,92 @@ class Submission {
...
+    async submissionAction(params, context){
+        verifySession(context)
+            .verifyInitialized();
+        const userInfo = context.userInfo;
+        const submissionID = params?.submissionID;
+        const action = params?.action;
+        //verify submission action
+        const verifier = verifySubmissionAction(submissionID, action);
+        //verify if a submission can be find by submissionID.
+        let submission = await verifier.exists(this.submissionCollection);
```
Comment (knockknockyoo) [inline services/submission.js:188]: I think we should not pass this.submissionCollection. The verifier should be separate, focusing only on verification.
const submission = await this.submissionCollection(id);
await verifier.exists(this.submissionCollection);

File: services/submission.js [hunk 7]
```diff
@@ -229,8 +170,92 @@ class Submission {
...
+        let fromStatus = submission.status;
+        //verify if the action is valid based on current submission status
+        verifier.isValidAction(submissionActionMap);
+        //verify if user's role is valide for the action
```
Comment (knockknockyoo) [inline services/submission.js:192]: valide => valid :)

**Review by knockknockyoo — CHANGES_REQUESTED** [review]

File: verifier/submission-verifier.js (hunk 1 above)
Comment (vshand11) [reply verifier/submission-verifier.js:11]: I will

File: verifier/submission-verifier.js (hunk 3 above)
Comment (vshand11) [reply verifier/submission-verifier.js:41]: I will

File: verifier/submission-verifier.js (hunk 4 above)
Comment (vshand11) [reply verifier/submission-verifier.js:34]: I will

File: verifier/submission-verifier.js (hunk 5 above)
Comment (vshand11) [reply verifier/submission-verifier.js:19]: good catch

File: services/submission.js (hunk 6 above)
Comment (vshand11) [reply services/submission.js:188]: I will consider, but my implement is a one stone server birds solution that encapsulate logics to verify the submissionId and check if submission can be found.

File: services/submission.js (hunk 7 above)
Comment (vshand11) [reply services/submission.js:192]: I will

**Comment by vshand11** [comment]
Updated, please a review.

**Review by knockknockyoo — APPROVED** [review]

**Review by AustinSMueller — APPROVED** [review]

**Pull request merged**